world/petitions/models.py

```diff
@@ -10,6 +10,93 @@
+    wantedads = models.ManyToManyField("WantedAd", related_name="wanted_ads")
+
+
+class WantedAd(SharedMemoryModel):
+    """This is an advertisement for wanting an RP partner"""
+    MARRIAGE = 0
+    SPONSOR = 1
+    PROTEGE = 2
+    PLOT = 3
+    TIME = 4
+    MATCH_TYPES = ((MARRIAGE, "Marriage"), (SPONSOR, "Sponsor"),
+                   (PROTEGE, "Protege"), (PLOT, "Plot"), (TIME, "Time"))
+    playtime_start = models.PositiveIntegerField(default=0)
+    playtime_end = models.PositiveIntegerField(default=0)
+    prefer_weekend = models.BooleanField(default=False)
+    interests = models.ManyToManyField(Interest, blank=True)
+    clues = models.ManyToManyField(Clue, blank=True)
+    searchtags = models.ManyToManyField(SearchTag, blank=True)
+    blurb = models.TextField()
+    marriage = models.BooleanField(default=False)
+    sponsor = models.BooleanField(default=False)
+    protege = models.BooleanField(default=False)
+    plot = models.BooleanField(default=False)
+    active = models.BooleanField(default=False)
+    owner = models.OneToOneField(
+        PlayerOrNpc,
+        related_name="wanted_ad",
+        on_delete=models.CASCADE)
+
+    def match(self, dompc):
+        matches = []
+
+        overlap = 0
+        try:
+            seeker_ad = dompc.wanted_ad
+        except AttributeError:
+            return matches
+        my_end = int(self.playtime_end)
+        their_end = int(seeker_ad.playtime_end)
+        my_start = int(self.playtime_start)
+        their_start = int(seeker_ad.playtime_start)
+
+        if dompc == self.owner:
+            return matches
+        if my_start >= my_end:
+            my_end += 24
+        if their_start >= their_end:
+            their_end += 24
+
+        if their_end > my_end:
+            overlap = my_end
+        else:
+            overlap = their_end
+
+        if their_start > my_start:
+            overlap -= their_start
+        else:
+            overlap -= my_start
+
+        if overlap > 2:
+            matches.append(self.TIME)
+        if self.sponsor and seeker_ad.protege:
+
+            rankcheck, diff = CmdPatronage().check_social_rank_difference(
+                self.owner.player.char_ob, dompc.player.char_ob)
+            if rankcheck:
+                matches.append(self.PROTEGE)
+        if self.protege and seeker_ad.sponsor:
+            rankcheck, diff = CmdPatronage().check_social_rank_difference(
+                dompc.player.char_ob, self.owner.player.char_ob)
+            if rankcheck:
+                matches.append(self.SPONSOR)
+        if self.marriage and seeker_ad.marriage:
+            matches.append(self.MARRIAGE)
+        if self.plot and seeker_ad.plot:
+            intersect_clues = self.clues.all().intersection(seeker_ad.clues.all())
+            intersect_tags = self.searchtags.all().intersection(seeker_ad.searchtags.all())
+            if intersect_clues or intersect_tags:
+                matches.append(self.PLOT)
+        return matches
```

Comment:
This could be very, very expensive if it was called in a loop. Ideally the match() should be replaced by a single query for matching clues and search tags if the one doing the query has plot up. I do django stuff in my day job so I don't mind handling this if it's not something you want to mess with.

Reply:
Didn't we already swap to using a single query for matching the clues and search tags? Match() currently only does everything else.